Handle additional live allocations not in checkpointed state

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96433
* #96384
* #96154
* #89146
* #95020
* __->__ #94943
* #94653

We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.

Differential Revision: [D43999886](https://our.internmc.facebook.com/intern/diff/D43999886)

eellison committed:
Handle additional live allocations not in checkpointed state

[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"




[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"




[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"




[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

zdevito commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -1045,9 +1048,15 @@ static void registerCudaPluggableAllocator(PyObject* module) {
       "_cuda_setCheckpointPoolState",
       [](int device,
-         std::shared_ptr<c10::cuda::CUDACachingAllocator::AllocatorState> pps) {
+         std::shared_ptr<c10::cuda::CUDACachingAllocator::AllocatorState> pps,
```
Perhaps it would make more sense for setCheckpointPoolState to return `allocated_stale_pointers` and then have this code here do the loop that disables their deleters. This way the allocator doesn't end up StorageImpl objects which are not part of its typical abstraction. It would also help simplify what future pluggable allocators would have to do to checkpoint and restore state. Basically the contract for setCheckpointPoolState will be that it obliterates the state of the allocator and returns a list of live pointers whose deleters need to be handled.

eellison replied:
Sounds good

eellison replied:
I made these changes.. 

The one issue here is that now since the storage swapping happens outside of the allocator, there is no way for the allocator to check that the "definitely freed" tensors have storages to swap onto. It will still throw an error in Module.cpp, but then there will be a second error when that tensor dies. Obviously we don't expect to hit this error case, but not ideal.

eellison replied:
I guess I would probably deliberating on fix in follow up but up to you. For now, I am the only user, and I am checking to make sure that I pass in the right tensors to deallocate.

zdevito replied:
I think for now this is ok, and when the APIs are all in place we can ask what series of actions can lead to this state and figure out the right guards then.

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

ezyang commented on test/test_cuda.py:
```diff
@@ -5245,11 +5245,10 @@ def test_allocated_in_middle_of_segment(self):
         def foo():
             small_buffers = [int8_cuda(MIN_BLOCK_SIZE) for _ in range(11)]
-            return small_buffers[5]
+            return small_buffers[5].add_(2)
```
hmm?

eellison replied:
will revert

ezyang commented on test/test_cuda.py:
```diff
@@ -5265,6 +5264,89 @@ def foo():
+        gc.collect()
+        torch.cuda.empty_cache()
+
+    # TODO: re-enable
```
You gonna reenable?

eellison replied:
There is another comment thread that I commented on this about. To enable this we would have to league Storages as a concept to the allocator, so i'm not sure. Explanation here: https://github.com/pytorch/pytorch/pull/94943#discussion_r1120749709.

ezyang replied:
ok, just breadcrumb it in the comment here, so we can find it later

ezyang commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -1045,9 +1071,45 @@ static void registerCudaPluggableAllocator(PyObject* module) {
+          freed_pointer_set.insert((ptr));
+        }
+        // that block has already been freed,
+        // so even those this will error, so too will the allcoator
```
allocator

apivovarov replied:
- allcoator
- correponding
- even those this
- so too will the allcoator

ezyang commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -1045,9 +1071,45 @@ static void registerCudaPluggableAllocator(PyObject* module) {
+        // when the corresponding tensor dies because there is no
+        // live tensor correponding to it
+        TORCH_CHECK(
+            ptr_set.size() >= definite_freed_count,
```
Is this a >=, or is it an exact equality test? Why is this an inequality?

eellison replied:
It could be larger in the case that there is a Block which is removed and freed with the same pointer address.

ezyang commented on c10/cuda/CUDACachingAllocator.h:
```diff
@@ -175,6 +177,14 @@ struct SnapshotInfo {
 
+// returns the pointers freed in the pool
+// and the pointers allocated. Note: a pointer
+// may appear in both freed and allocated
```
So what does it mean if a pointer appears in both freed and allocated? Is it live coming out or not?

eellison replied:
It could be reallocated as a separate size.

ezyang replied:
Can I have the opposite situation? A tensor is allocated and then freed internally?

eellison replied:
Yes that is possible. We pass in the stale storages to the api and them swap their deleters with no-ops so they are not freed twice.

ezyang commented on torch/csrc/cuda/Module.cpp:
```diff
@@ -917,6 +920,29 @@ static void registerCudaDeviceProperties(PyObject* module) {
+    bool succeeded = stale_storage->data_ptr().compare_exchange_deleter(
+        t->raw_deleter(), &no_op_delete);
+
+    TORCH_CHECK(
```
So, when can this happen? Is this if I pass garbage tensors to the checkpoint function? If I only ever pass tensors that are allocated out of the pool, the deleter should always be as expected, right? Can we describe the preconditions for the checkpoint tensor list?

eellison replied:
This error means that you had a non-standard deleter fn, and so it's not safe for us to blindly remove it. As the comment in Allocator.h says:

```
   * Compare the deleter in a DataPtr to expected_deleter.
   * If it matches, replace the deleter with new_deleter
   * and return true; otherwise, does nothing and returns
   * false.
   *
   * In general, it is not safe to unconditionally set the
   * deleter on a DataPtr, because you don't know what
   * the deleter is, and thus will have a hard time properly
   * disposing of the deleter without storing the original
   * deleter
   ```

zdevito commented on c10/cuda/CUDACachingAllocator.cpp:
```diff
@@ -110,6 +110,8 @@ constexpr size_t kMinLargeAlloc =
 constexpr size_t kRoundLarge = 2097152; // round up large allocations to 2 MiB
 constexpr size_t kRoundUpPowerOfTwoIntervals = 16;
 
+void local_raw_delete(void* ptr);
```
nit: don't think this is needed anymore, though it isn't doing any harm to be pre-declared...

zdevito reviewed (approved):
Looks nice and clean now. Thanks for the changes!

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.


[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999886).

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.

Differential Revision: [D43999886](https://our.internmc.facebook.com/intern/diff/D43999886)

[ghstack-poisoned]

eellison committed:
Update on "Handle additional live allocations not in checkpointed state"


We choose to ignore certain blocks that are currently allocated when we set the pool to its checkpoint. For those blocks, we need to swap out the deleter function of their corresponding blocks so that a deallocation is not triggered when they die.

Differential Revision: [D43999886](https://our.internmc.facebook.com/intern/diff/D43999886)

[ghstack-poisoned]

eellison commented:
@pytorchbot merge

apivovarov commented:
@eellison Can you look again at the comments in Module.cpp. It has the following typos/mistakes:
- allcoator
- correponding
- even those this
- so too will the allcoator

eellison commented:
Will accept a pr to fix.

apivovarov commented:
Should we replace
_so even those this will error, so too will the allocator_
with
_Although this will result in an error, the allocator will also encounter the same issue._
?

eellison commented:
sure, that works.
